modules/javafx.base/src/main/java/com/sun/javafx/collections/ConcatenatedObservableList.java

```diff
@@ -0,0 +1,161 @@
+
+    private void onListChanged(ListChangeListener.Change<? extends E> change) {
+        int listOffset = computeListOffset(change.getList());
+
+        while (change.next()) {
+            beginChange();
+
+            if (change.wasPermutated()) {
+                onPermutated(change, listOffset);
+            } else if (change.wasUpdated()) {
+                onUpdated(change, listOffset);
+            } else {
+                if (change.wasReplaced()) {
+                    onReplaced(change, listOffset);
+                } else if (change.wasRemoved()) {
+                    onRemoved(change, listOffset);
+                } else if (change.wasAdded()) {
+                    onAdded(change, listOffset);
+                }
+            }
+
+            endChange();
+        }
+    }
+
+    private void onPermutated(ListChangeListener.Change<? extends E> change, int listOffset) {
+        int from = change.getFrom();
+        int to = change.getTo();
+
+        if (listOffset == 0 && change instanceof NonIterableChange.SimplePermutationChange<?> permChange) {
+            nextPermutation(from, to, permChange.getPermutation());
+        } else {
+            int[] perm = new int[to - from];
+
+            for (int i = 0, oldIndex = change.getFrom(), max = change.getTo(); oldIndex < max; ++i, ++oldIndex) {
+                int newIndex = change.getPermutation(oldIndex);
+                perm[i] = newIndex + listOffset;
+            }
+
+            nextPermutation(from, to, perm);
+        }
+    }
+
+    private void onUpdated(ListChangeListener.Change<? extends E> change, int listOffset) {
+        for (int index = change.getFrom(), max = change.getTo(); index < max; ++index) {
+            nextUpdate(listOffset + index);
+        }
+    }
+
+    private void onReplaced(ListChangeListener.Change<? extends E> change, int listOffset) {
+        int from = change.getFrom(), to = change.getTo();
+        size = addSize(to - from - change.getRemovedSize());
+        nextReplace(listOffset + from, listOffset + to, change.getRemoved());
+    }
+
+    private void onRemoved(ListChangeListener.Change<? extends E> change, int listOffset) {
+        size -= change.getRemovedSize();
+        nextRemove(change.getFrom() + listOffset, change.getRemoved());
+    }
+
+    private void onAdded(ListChangeListener.Change<? extends E> change, int listOffset) {
+        int from = change.getFrom(), to = change.getTo();
+        size = addSize(to - from);
+        nextAdd(from + listOffset, to + listOffset);
+    }
+
+    @Override
+    public E get(int index) {
+        Objects.checkIndex(index, size);
+        int newIndex = index;
+
+        for (ObservableList<? extends E> list : lists) {
+            int listSize = list.size();
+            if (newIndex >= listSize) {
+                newIndex -= listSize;
+            } else {
+                return list.get(newIndex);
+            }
+        }
+
+        throw new IndexOutOfBoundsException(index);
```

Comment:
I think this should be an `AssertionError`, it can't be covered in a unit test as `checkIndex` prevents this case from ever happening.